### ISSUE TITLE
Publish to different channels using only one WebSocket connection

### ISSUE THREAD

**Issue by frelars** [slact/nchan#425]
I have a usecase where clients (many) establish and listen to "private" channel so services/producers are able to send messages to individual clients. The message rate is high so we use websockets today.

As far as I can tell, nchan requires publishers to establish a new websocket connection for every subscriber it want to talk to, which will be very inefficient for our usecase. Please correct me if I'm wrong.

Is there any plan for supporting dynamic channel destinations for websocket publishers? I'm considering adding support for this my self as the code seems very readable and the publisher logic seems manageable, but I may very well be wrong as I only spent 10 min browsing the code :)

In my case I'm considering reserving the first line of every websocket frames routing info (channel destination). The initial connect will just be to a random (unused) channel, as a channel id is needed to connect. The real channel destination will be picked up in the message body itself. This will probably destroy binary support.

If you already have plans for something similar it would be nice to know. If not it would be nice to get some input as to how realistic above plan seems

**Comment by slact** [comment]
> I have a usecase where clients (many) establish and listen to "private" channels so services/producers are able to send messages to individual clients. The message rate is high so we use websockets today.

 For publishing to different channels, the best protocol is http2. It keeps the TCP connection open and has virtually no headers overhead. HTTP 1.1 work well too, as the TCP connection there is also keep-alive by default, however, the headers overhead is higher.

> As far as I can tell, nchan requires publishers to establish a new websocket connection for every subscriber it wants to talk to

Correct. There is currently no way to change  channels for  a publisher/subscriber.

>  Is there any plan for supporting dynamic channel destinations for websocket publishers

Yes, this is on the TODO list, implemented as special control messages (basically encapsulated  HTTP requests).

> In my case I'm considering reserving the first line of every websocket frame for routing info (channel destination).

I recommend going with the http2 route for this. Mind your premature optimizations, and benchmark publishing performance first. If it's not enough, then start thinking about custom  websocket subprotocols.

**Comment by frelars** [comment]
Thanks for the feedback slact! Yes, I'm well aware of http2 header compression+multiplexing, keep-alive (2 and 1.1) and the associated benefits. We already depend on that for our client side (+ws). But on serverside we use websockets for all the things (micro services, connected to a ws-hub) and it would be very risky to change all internal com infrastructure over to http2 just to use chan.

If nchan supported websocket and a protocol for publishing to dynamic destinations, we could replicate our internal hub topology without changing much else. This would also allow us to drop some internal services and also gain some performance and stability (probably). In this case the little work required would be well worth it.

Anyway, I understand from you response that patching nchan to allow what I want will not be trivial, so I will drop this idea. I think however that this is a important feature so hopefully your todo list is not to long :) Anyway, great project! One day I will be a user (and possible a contributor)

**Comment by slact** [comment]
Ok, that makes sense. 

Is this something you're willing to sponsor? Shoot me an email at leo@slact.net if you're interested.

**Comment by frelars** [comment]
If implementing this functionality allows us to optimise our infrastructure then sponsoring such a development make sense to me. But this assumes everything else in nchan is production ready (battle tested) so we don´t need to (once again..) spend 3 months debugging strange net/service behaviour.

For now we are just reviewing possible improvements to our infrastructure and nchan is very interesting candidate. When we decide to improve stuff we will consider sponsoring discussed feature.

**Comment by slact** [comment]
Nchan has been in production use for a few years now, and the push module, its predecessor, for even longer. Bugs still exist, of course, but I fix them rather quickly whenever they come up.

Do let me know what you end up deciding -- I already have someone else interested in sponsoring this feature, but it's not quite enough to get the work finished. Your contribution can make the difference.

**Comment by frelars** [comment]
I know nchan have seen some production use, but given my feature request I don´t think many other companies use it as we plan/want to use it. So it is still an open question if we can fit it to our environment and how good fit it will be even if dynamic ws channels are supported.

For instance, every client today only opens 1 ws to hub. It uses this one socket to emit queries to different services. Responses are then pushed back asynchronously to the client using the same channel. This will also require some changes, either in nchan or how we do stuff. It is doable and I suspect nchan can do this very good, but it will require resources and it is not a direct value driver for us, so hard to prioritise during busy days.

Anyway, I will contact you if/when we start getting serious about changing infrastructure. I think the result could be very good, but it will be costly in terms of time and money.

**Comment by dungvu-qe** [comment]
I've faced the same issue, what about the feature? maybe a year from the last comment, any update for this?

**Comment by majidbha** [comment]
I've faced the same issue, what about the feature? maybe a year from the last comment, any update for this?

**Comment by slact** [comment]
Yes, this feature is going to be implemented in the coming months, as well as quite a lot more. ETA sometime this summer.
